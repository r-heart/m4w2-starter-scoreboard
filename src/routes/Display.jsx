import PropTypes from "prop-types";
import { useEffect, useState } from "react";
import Button from "../components/Buttons/Button.jsx";
import Buttons from "../components/Buttons/Buttons.jsx";
import { HomeAwaySwitch } from "../components/Form";
import Scoreboard from "../components/Scoreboard";

export default function Display({ buttons, periods, timePerPeriod }) {
  const [homeAway, setHomeAway] = useState("away");
  const [awayScore, setAwayScore] = useState(0);
  const [homeScore, setHomeScore] = useState(0);
  const [currentPeriod, setCurrentPeriod] = useState(1);
  const [timeRemaining, setTimeRemaining] = useState(
    // Seconds to minutes
    timePerPeriod * 60
  );
  const [isTimerRunning, setIsTimerRunning] = useState(false);

  useEffect(() => {
    if (isTimerRunning && timeRemaining > 0) {
      const intervalId = setInterval(() => {
        setTimeRemaining((prev) => prev - 1);
      }, 1000);

      return () => {
        clearInterval(intervalId);
      };
    }
  }, [timeRemaining, isTimerRunning]);

  return (
    <>
      <HomeAwaySwitch
        handleToggle={() => {
          setHomeAway((prev) => (prev === "home" ? "away" : "home"));
        }}
      />

      <Buttons
        buttons={buttons}
        handleClick={(e) => {
          if (homeAway === "home") {
            setHomeScore((prev) => prev + Number(e.target.dataset.count));
          } else {
            setAwayScore((prev) => prev + Number(e.target.dataset.count));
          }
        }}
      />

      <Scoreboard
        homeScore={homeScore}
        awayScore={awayScore}
        timeRemaining={timeRemaining}
        currentPeriod={currentPeriod}
      />

      <div className="flex gap-x-8">
        <Button
          colorClass="bg-green-500"
          text="Start"
          handleClick={() => {
            setIsTimerRunning((prev) => !prev);
          }}
        />

        <Button
          colorClass="bg-orange-500"
          text="Stop"
          handleClick={() => {
            setIsTimerRunning((prev) => !prev);
          }}
        />

        <Button
          colorClass="bg-yellow-700"
          text="Next Period"
          handleClick={() => {
            console.log("Next Period");
          }}
        />

        <Button
          colorClass="bg-red-500"
          text="Reset Board"
          handleClick={() => {
            console.log("Reset Board");
          }}
        />
      </div>
    </>
  );
}

Display.propTypes = {
  buttons: PropTypes.arrayOf(PropTypes.number).isRequired,
  periods: PropTypes.oneOfType([PropTypes.number, PropTypes.string]).isRequired,
  timePerPeriod: PropTypes.oneOfType([PropTypes.number, PropTypes.string])
    .isRequired,
};
