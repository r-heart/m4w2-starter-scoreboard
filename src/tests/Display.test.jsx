import { act, render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { vi } from "vitest";
import Display from "../routes/Display";

test("initial render", () => {
  const rendered = render(
    <Display buttons={[1, 2, 3]} periods={3} timePerPeriod={5} />
  );
  expect(rendered).toMatchSnapshot();
});

it("updates the Away score", () => {
  it("updates the Away score", async () => {
    const user = userEvent.setup();
    render(<Display buttons={[1, 2, 3]} periods={3} timePerPeriod={5} />);
    const awayButton = screen.getByRole("button", { name: "1" });
    const awayScore = screen.getByTestId("away-score");
    await user.click(awayButton);
    expect(awayScore).toHaveTextContent("1");
  });
});

it("update the Home and Away score based on the toggles", async () => {
  const user = userEvent.setup();
  render(<Display buttons={[1, 2, 3]} periods={3} timePerPeriod={5} />);
  const buttons = screen.getAllByRole("button");

  // Toggle starts with "away" selected
  const toggle = screen.getByRole("checkbox");

  const awayScore = screen.getByTestId("away-score");
  const homeScore = screen.getByTestId("home-score");

  await user.click(buttons[0]);

  expect(awayScore).toHaveTextContent("1");
  expect(homeScore).toHaveTextContent("0");

  await user.click(toggle);
  await user.click(buttons[0]);

  expect(awayScore).toHaveTextContent("1");
  expect(homeScore).toHaveTextContent("1");

  await user.click(toggle);
  await user.click(buttons[0]);

  expect(awayScore).toHaveTextContent("2");
  expect(homeScore).toHaveTextContent("1");
});

it("advances the period only up to the max periods", async () => {
  const user = userEvent.setup();
  render(<Display buttons={[1, 2, 3]} periods={2} timePerPeriod={5} />);

  const nextPeriodButton = screen.getByRole("button", { name: "Next Period" });
  const periodP = screen.getByTestId("period");

  await user.click(nextPeriodButton);
  expect(periodP).toHaveTextContent("2");

  // Limited to two periods
  await user.click(nextPeriodButton);
  expect(periodP).toHaveTextContent("2");
});

// Testing Timers
describe("Timer 🤡", () => {
  beforeAll(() => {
    vi.useFakeTimers();
  });

  test("timer reflects MM:SS accurately", async () => {
    const user = userEvent.setup({ delay: null });
    render(<Display buttons={[1, 2, 3]} periods={2} timePerPeriod={10} />);

    const startBtn = screen.getByRole("button", { name: "Start" });
    const timeDisplay = screen.getByTestId("time");

    await user.click(startBtn);

    await waitFor(() => {
      expect(timeDisplay).toHaveTextContent("10:00");
    });

    act(() => {
      // 30 seconds
      vi.advanceTimersByTime(30000);
    });

    await waitFor(() => {
      expect(timeDisplay).toHaveTextContent("9:30");
    });

    act(() => {
      // 6 more seconds
      vi.advanceTimersByTime(6000);
    });

    await waitFor(() => {
      expect(timeDisplay).toHaveTextContent("9:24");
    });

    act(() => {
      // 1 second left! ⏳
      vi.advanceTimersByTime(563000);
    });

    await waitFor(() => {
      expect(timeDisplay).toHaveTextContent("0:01");
    });
  });

  test("stops and restarts timer", async () => {
    const user = userEvent.setup({ delay: null });
    render(<Display buttons={[1, 2, 3]} periods={2} timePerPeriod={10} />);

    const startBtn = screen.getByRole("button", { name: "Start" });
    const stopBtn = screen.getByRole("button", { name: "Stop" });
    const timeDisplay = screen.getByTestId("time");

    await user.click(startBtn);

    act(() => {
      // 30 seconds
      vi.advanceTimerByTime(30000);
    });

    await user.click(stopBtn);

    act(() => {
      vi.advanceTimersByTime(10000);
    });

    await waitFor(() => {
      // Even though 40 seconds have elapsed...
      expect(timeDisplay).toHaveTextContent("9:30");
    });

    await user.click(startBtn);

    act(() => {
      vi.advanceTimersByTime(10000);
    });

    await waitFor(() => {
      expect(timeDisplay).toHaveTextContent("9:20");
    });
  });

  test("resets the time display when period is advanced", async () => {
    const user = userEvent.setup({ delay: null });
    render(<Display buttons={[1, 2, 3]} periods={2} timePerPeriod={10} />);

    const startBtn = screen.getByRole("button", { name: "Start" });
    const nextPeriodBtn = screen.getByRole("button", { name: "Next Period" });
    const timeDisplay = screen.getByTestId("time");

    await user.click(startBtn);

    act(() => {
      // 30 seconds
      vi.advanceTimersByTime(30000);
    });

    await waitFor(() => {
      expect(timeDisplay).toHaveTextContent("9:30");
    });

    await user.click(nextPeriodBtn);

    await waitFor(() => {
      expect(timeDisplay).toHaveTextContent("10:00");
    });
  });
});
