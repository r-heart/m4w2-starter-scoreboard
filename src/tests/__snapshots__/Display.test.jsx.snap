// Vitest Snapshot v1

exports[`initial render 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <label
        class="switch relative flex cursor-pointer items-center"
        for="switch"
      >
        <span
          class="mr-1 text-sm"
          data-testid="home-toggle"
        >
          Home
        </span>
        <input
          checked=""
          class="sr-only"
          id="switch"
          type="checkbox"
        />
        <span
          class="h-6 w-12 rounded-full bg-gray-800"
        />
        <span
          class="ml-1 text-sm"
          data-testid="away-toggle"
        >
          Away
        </span>
      </label>
      <div
        class="flex flex-row gap-x-4"
      >
        <button
          class="rounded bg-blue-500 px-4 py-2 text-white"
          data-count="1"
        >
          1
        </button>
        <button
          class="rounded bg-blue-500 px-4 py-2 text-white"
          data-count="2"
        >
          2
        </button>
        <button
          class="rounded bg-blue-500 px-4 py-2 text-white"
          data-count="3"
        >
          3
        </button>
      </div>
      <section
        class="flex flex-row gap-x-8"
      >
        <div
          class="flex flex-col items-center"
        >
          <h2
            class="text-2xl"
          >
            Home
          </h2>
          <p
            class="text-4xl"
            data-testid="home-score"
          >
            0
          </p>
        </div>
        <div
          class="flex flex-col items-center gap-y-4"
        >
          <div
            class="flex flex-col items-center"
          >
            <h2
              class="text-2xl"
            >
              Period
            </h2>
            <p
              class="text-4xl"
              data-testid="period"
            >
              1
            </p>
          </div>
          <div
            class="flex flex-col items-center"
          >
            <h2
              class="text-2xl"
            >
              Time
            </h2>
            <p
              class="text-4xl"
              data-testid="time"
            >
              5:00
            </p>
          </div>
        </div>
        <div
          class="flex flex-col items-center"
        >
          <h2
            class="text-2xl"
          >
            Away
          </h2>
          <p
            class="text-4xl"
            data-testid="away-score"
          >
            0
          </p>
        </div>
      </section>
      <div
        class="flex gap-x-8"
      >
        <button
          class="rounded bg-green-500 px-4 py-2 text-white"
          data-count="Start"
        >
          Start
        </button>
        <button
          class="rounded bg-orange-500 px-4 py-2 text-white"
          data-count="Stop"
        >
          Stop
        </button>
        <button
          class="rounded bg-yellow-700 px-4 py-2 text-white"
          data-count="Next Period"
        >
          Next Period
        </button>
        <button
          class="rounded bg-red-500 px-4 py-2 text-white"
          data-count="Reset Board"
        >
          Reset Board
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <label
      class="switch relative flex cursor-pointer items-center"
      for="switch"
    >
      <span
        class="mr-1 text-sm"
        data-testid="home-toggle"
      >
        Home
      </span>
      <input
        checked=""
        class="sr-only"
        id="switch"
        type="checkbox"
      />
      <span
        class="h-6 w-12 rounded-full bg-gray-800"
      />
      <span
        class="ml-1 text-sm"
        data-testid="away-toggle"
      >
        Away
      </span>
    </label>
    <div
      class="flex flex-row gap-x-4"
    >
      <button
        class="rounded bg-blue-500 px-4 py-2 text-white"
        data-count="1"
      >
        1
      </button>
      <button
        class="rounded bg-blue-500 px-4 py-2 text-white"
        data-count="2"
      >
        2
      </button>
      <button
        class="rounded bg-blue-500 px-4 py-2 text-white"
        data-count="3"
      >
        3
      </button>
    </div>
    <section
      class="flex flex-row gap-x-8"
    >
      <div
        class="flex flex-col items-center"
      >
        <h2
          class="text-2xl"
        >
          Home
        </h2>
        <p
          class="text-4xl"
          data-testid="home-score"
        >
          0
        </p>
      </div>
      <div
        class="flex flex-col items-center gap-y-4"
      >
        <div
          class="flex flex-col items-center"
        >
          <h2
            class="text-2xl"
          >
            Period
          </h2>
          <p
            class="text-4xl"
            data-testid="period"
          >
            1
          </p>
        </div>
        <div
          class="flex flex-col items-center"
        >
          <h2
            class="text-2xl"
          >
            Time
          </h2>
          <p
            class="text-4xl"
            data-testid="time"
          >
            5:00
          </p>
        </div>
      </div>
      <div
        class="flex flex-col items-center"
      >
        <h2
          class="text-2xl"
        >
          Away
        </h2>
        <p
          class="text-4xl"
          data-testid="away-score"
        >
          0
        </p>
      </div>
    </section>
    <div
      class="flex gap-x-8"
    >
      <button
        class="rounded bg-green-500 px-4 py-2 text-white"
        data-count="Start"
      >
        Start
      </button>
      <button
        class="rounded bg-orange-500 px-4 py-2 text-white"
        data-count="Stop"
      >
        Stop
      </button>
      <button
        class="rounded bg-yellow-700 px-4 py-2 text-white"
        data-count="Next Period"
      >
        Next Period
      </button>
      <button
        class="rounded bg-red-500 px-4 py-2 text-white"
        data-count="Reset Board"
      >
        Reset Board
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
